XIPs/xip-49-decentralized-backend.md

```diff
@@ -0,0 +1,1099 @@
+
+```protobuf
+
+// Request to retrieve the XIDs for the given addresses
+message GetInboxIdsRequest {
+  // A single request for a given address
+  message Request {
+    string address = 1;
+  }
+  repeated Request requests = 1;
+}
+
+// Response with the XIDs for the requested addresses
+message GetInboxIdsResponse {
+  // A single response for a given address
+  message Response {
+    string address = 1;
+    optional string inbox_id = 2;
+  }
+  repeated Response responses = 1;
+}
+
+service ReplicationApi {
+  rpc GetInboxIds(GetInboxIdsRequest) returns (GetInboxIdsResponse) {
+    option (google.api.http) = {
+      post: "/mls/v2/get-inbox-ids"
+      body: "*"
+    };
+  }
+}
+```
+
+#### 3.3.4. Validation rules
+
+On message publish, originators must validate the incoming `PayerEnvelope`:
+
+1. Verify payer envelope signature
+2. Verify client envelope AAD:
+    1. Target originator matches self
+    2. Target topic matches payload type, as defined in [3.3.2. Envelopes](#332-envelopes)
+    3. The server’s cursor is greater than `last_seen`. If not, a 409 status code must be returned with the server’s current cursor.
+    4. The blockchain sequence ID in the cursor is equal to the latest blockchain payload published on the topic. If not, a 409 status code must be returned with the server’s current cursor.
+3. Perform payload-specific validation, as defined in previous XIP’s.
+4. Publish commits or identity updates to the blockchain, else sign the payload and originate it.
+
+On message publish, clients must validate the returned `OriginatorEnvelope`:
+
+1. Verify that the enclosed `PayerEnvelope` matches the originally published `PayerEnvelope`.
+2. Verify the `originator_signature` or `node_signature`.
+3. Verify `originator_ns` is within 30 mins of the client clock.
+4. (Optional) Query a peer node, with an exponential back-off, for the returned sequence ID to ensure it was replicated.
+
+In the event that these rules are broken, clients may publish a misbehavior report and retry against another originator.
+
+On message query, in addition to validation required by MLS, clients must:
+
+1. Verify the `originator_sequence_id` is strictly increasing per originator.
+2. Verify the `originator_signature` or `node_signature` on each payload.
+3. Verify the client envelope AAD on each payload:
+    1. The target originator matches the actual originator.
+    2. The target topic matches the actual topic.
+4. Verify the client envelope AAD was correctly authenticated ([3.3.5. Cross-originator message ordering](#335-cross-originator-message-ordering))
+5. Verify that the payloads are causally consistent ([3.3.5. Cross-originator message ordering](#335-cross-originator-message-ordering)).
+6. Perform payload-specific validation, as defined in section XXX.
```

Comment:
@richardhuaaa - where should we link this? TY!

Reply:
Good catch! I think this is sufficient

```suggestion
6. Perform payload-specific validation, as defined in previous XIP's.
```